pkg/operator/ceph/cluster/status/conditions.go

```diff
@@ -0,0 +1,300 @@
+		applyCondition(context, conditions, newCondition)
+		return
+	}
+	if existingCondition.Status != newCondition.Status {
+		removeStatusCondition(conditions, newCondition.Type)
+		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		applyCondition(context, conditions, newCondition)
+	}
+}
+
+func ConditionConnecting(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionConnecting,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionConnected(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionConnected,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionIgnored(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionIgnored,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionUpgrading(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionNotReady,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionReady,
+		Status:  v1.ConditionFalse,
+		Reason:  "",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionAvailable,
+		Status:  v1.ConditionFalse,
+		Reason:  "",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionUpgrading,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionFailure(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionReady,
+		Status:  v1.ConditionFalse,
+		Reason:  "ClusterNotReady",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionAvailable,
+		Status:  v1.ConditionFalse,
+		Reason:  "",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionNotReady,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionFailure,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionProgressing(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionNotReady,
```

Comment:
Will ConditionNotReady be set to true every time an orchestration is run? If the cluster was initially created successfully, the cluster is hopefully still in a ready state even while the upgrade is progressing.